src/fdc3/dotnet/DesktopAgent.Client/src/MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client/Infrastructure/DesktopAgentClient.cs

```diff
@@ -148,14 +166,67 @@ public Task<IEnumerable<IChannel>> GetUserChannels()
         throw new NotImplementedException();
     }
 
-    public Task JoinUserChannel(string channelId)
+    public async Task JoinUserChannel(string channelId)
     {
-        throw new NotImplementedException();
+        try
+        {
+            await _currentChannelLock.WaitAsync().ConfigureAwait(false);
+
+            if (_currentChannel != null)
+            {
+                _logger.LogInformation($"Leaving current channel: {_currentChannel.Id}...");
+                await LeaveCurrentChannel().ConfigureAwait(false);
+            }
+
+            if (!_userChannels.TryGetValue(channelId, out var channel))
+            {
+                channel = await _channelFactory.JoinUserChannelAsync(channelId).ConfigureAwait(false);
+                _userChannels[channelId] = channel;
+            }
+
+            _currentChannel = channel;
+        }
+        finally
+        {
+            var contextListeners = _contextListenersWithSubscriptionLastContextHandlingActions.Reverse();
```

Comment:
Seems a bit dangerous to put this much logic in the finally block and release the lock at the end of the same.
LeaveCurrentChannel has the same logic still in the try part and it is being called when leaving the previous channel, this seems like an oversight.

Reply:
This part is responsible to handle the subscription for the newly joined user channel for each top-level context listeners. As they should remain in the cache for the entire App lifetime - if I remember correctly.

For the `LeaveCurrentChannel` it calls the cached top-level context listener to unsubscribe from the user channel.
Do you have any suggestion how this could be made simpler? I am happy to refactor it; I was just out of ideas. :D